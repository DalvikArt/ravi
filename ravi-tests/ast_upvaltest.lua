local str = 
[[return function()
    local a = 1

	local function y()
		return function()
			return a
		end
	end

	local a = 5

	local function z()
		return function()
			return a
		end
	end

	return y, z
end
]]

local x, msg = ast.parse(str)
if not x then print(msg) end
print(x:tostring())

str =
[[return function (a, b, c, d, e)
    if a == b then goto l1
    elseif a == c then goto l2
    elseif a == d then goto l2
    else if a == e then goto l3
        else goto l3
        end
    end
    ::l1:: ::l2:: ::l3:: ::l4::
end
]]

x, msg = ast.parse(str)
if not x then print(msg) end
print(x:tostring())

