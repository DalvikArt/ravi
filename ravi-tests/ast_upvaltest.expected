function()
  return
    function()
      local
      --[symbols]
        a --local symbol ? 
      --[expressions]
        1
      local
      --[symbols]
        y --local symbol function 
      --[expressions]
        function()
        --upvalues  a
          return
            function()
            --upvalues  a
              return
                --[primary start]
                  a --upvalue ? 
                --[primary end]
            end
        end
      local
      --[symbols]
        a --local symbol ? 
      --[expressions]
        5
      local
      --[symbols]
        z --local symbol function 
      --[expressions]
        function()
        --upvalues  a
          return
            function()
            --upvalues  a
              return
                --[primary start]
                  a --upvalue ? 
                --[primary end]
            end
        end
      return
        --[primary start]
          y --local symbol function 
        --[primary end]
       ,
        --[primary start]
          z --local symbol function 
        --[primary end]
    end
end

function()
  return
    function(
      a --local symbol ? 
     ,
      b --local symbol ? 
     ,
      c --local symbol ? 
     ,
      d --local symbol ? 
     ,
      e --local symbol ? 
    )
      if
       --[binary expr start]
        --[primary start]
          a --local symbol ? 
        --[primary end]
       ==
        --[primary start]
          b --local symbol ? 
        --[primary end]
       --[binary expr end]
      then
        goto l1
      elseif
       --[binary expr start]
        --[primary start]
          a --local symbol ? 
        --[primary end]
       ==
        --[primary start]
          c --local symbol ? 
        --[primary end]
       --[binary expr end]
      then
        goto l2
      elseif
       --[binary expr start]
        --[primary start]
          a --local symbol ? 
        --[primary end]
       ==
        --[primary start]
          d --local symbol ? 
        --[primary end]
       --[binary expr end]
      then
        goto l2
      else
        if
         --[binary expr start]
          --[primary start]
            a --local symbol ? 
          --[primary end]
         ==
          --[primary start]
            e --local symbol ? 
          --[primary end]
         --[binary expr end]
        then
          goto l3
        else
          goto l3
        end
      end
      ::l1::
      ::l2::
      ::l3::
      ::l4::
    end
end

